### PR TITLE
Temporal datatype support for interval arithmetic

### LINKED ISSUE #5844: Temporal datatype support for interval arithmetic
### Is your feature request related to a problem or challenge?

Currently, interval arithmetic operations are not supported for temporal datatypes. It is especially necessary to be able to apply symmetric hash join over ordered timestamp columns, probably with some interval differences. 

### Describe the solution you'd like

[Interval arithmetic](https://github.com/apache/arrow-datafusion/blob/main/datafusion/physical-expr/src/intervals/interval_aritmetic.rs) and [cp solver](https://github.com/apache/arrow-datafusion/blob/main/datafusion/physical-expr/src/intervals/cp_solver.rs) are modified to handle timestamp and interval types. In [datetime.rs](https://github.com/apache/arrow-datafusion/blob/main/datafusion/physical-expr/src/expressions/datetime.rs), there is an evalute_array function where elements of an array is computed from an array and a scalar value. The scope of that evaluation is extended.

### Describe alternatives you've considered

-

### Additional context

-

### PR DESCRIPTION
**Pull request by berkaysynnada**
# Which issue does this PR close?

Closes #5844.

# Rationale for this change

Currently, interval arithmetic operations are not supported for temporal datatypes. It is especially necessary to be able to apply symmetric hash join over ordered timestamp columns, probably with some interval differences. 

With this PR, we can make use of interval arithmetic with Interval-Interval and Timestamp-Timestamp intervals. Also, evalute_array function (where array vs scalar operations go to in datetime.rs) had a limited evaluation ability. It has a similar behavior now with [that feature ](https://github.com/apache/arrow-datafusion/pull/5764).

# What changes are included in this PR?

evaluate_array is extended to make timestamp vs timestamp, timestamp vs interval, and interval vs interval calculations, and dangerous unwraps are removed. min/max of different interval types can be found now. `evaluate_bounds` and `propagate_constraints` are implemented for `DateTimeIntervalExpr`.

# Are these changes tested?

Yes, symmetric hash join tests are extended with temporal datatypes.

# Are there any user-facing changes?

### CODE REVIEW

**Review by mustafasrepo — APPROVED** [review]
This PR passed from our internal review process. It is LGTM!

**Comment by mustafasrepo** [comment]
I plan to merge this PR, If there is no additional review from the community. if there is anyone want to review, please let me know.

File: datafusion/common/src/scalar.rs
Comment (alamb) [inline datafusion/common/src/scalar.rs:1195]: TIL `div_euclid`

**Review by alamb — COMMENTED** [review]
Thank you @berkaysynnada  and @mustafasrepo  -- I didn't review this PR fully but I think the structure looks good to me. 

My biggest concern in the entire exercise is that the arithmetic that is supported for ScalarValue and what is supported for Arrays (and what is supported upstream in arrow-rs) will diverge.  However I think we are all agreed on the desired end state (kernels in arrow-rs and the same math / support for scalars in Datafusion)

**Comment by berkaysynnada** [comment]
> Thank you @berkaysynnada and @mustafasrepo -- I didn't review this PR fully but I think the structure looks good to me.
> 
> My biggest concern in the entire exercise is that the arithmetic that is supported for ScalarValue and what is supported for Arrays (and what is supported upstream in arrow-rs) will diverge. However I think we are all agreed on the desired end state (kernels in arrow-rs and the same math / support for scalars in Datafusion)

Thank you for the review. Once you review and approve the [structure](https://github.com/synnada-ai/arrow-datafusion/pull/83), I will move the other arithmetic codes that are for Arrays, and finally, I believe we can unify all arithmetics in kernels_arrow.rs

**Pull request merged**
